engine/source_sym_engine.ml

```diff
@@ -0,0 +1,184 @@
+      bprintf buf
+         "Conjunction:%t%a"
+         (break (ntabs+1))
+         (bprint_list ~sep (bprint_tree (ntabs+1))) cst_list
+    | Root cst_list ->
+       bprintf buf
+         "Root=\
+          %t%a"
+         (break (ntabs+1))
+         (bprint_list ~sep (bprint_tree (ntabs+1))) cst_list
+  in
+  let buf = Buffer.create 42 in
+  bprint_tree 0 buf stree;
+  BatIO.write_line BatIO.stdout (Buffer.contents buf)
+
+let sym_exec (source: source_program) =
+  let shares_kst ?wildcard_to_none:(w=false) k (clause:clause): clause option =
+    let pattern, expr = clause in 
+    match pattern with
+    | Wildcard when w = false-> Some (Wildcard, expr)
+    | Constructor (k', plist) when k' = k ->
+      if (List.length plist > 1) then
+        Some (Constructor (Tuple, plist), expr)
+      else
+        Some (plist |> List.hd, expr)
+    | _ -> None
+  in
+  let negate_piop = function
+    | EqInt i -> NeqInt i
+    | NeqInt i -> EqInt i
+    | IsK s -> NotK s
+    | NotK s -> IsK s
+    | IsTrue -> IsFalse
+    | IsFalse -> IsTrue
+    | IsTuple -> NotTuple
+    | NotTuple -> IsTuple
+    | IsNil -> NotNil
+    | NotNil -> IsNil
+    | IsCons -> NotCons
+    | NotCons -> IsCons
+    | EqString s -> NeqString s
+    | NeqString s -> EqString s
+  in
+  let piop_of_constructor = function
+    | Variant s -> IsK s
+    | Int i -> EqInt i
+    | Bool b -> if b then IsTrue else IsFalse
+    | String s -> EqString s
+    | Tuple -> IsTuple
+    | Nil -> IsNil
+    | Cons -> IsCons
+  in
+  let rec eval_clauses ?idx:(idx=0) constraints clauses : constraint_tree list =
+    match clauses with
+    | clause::tl -> let pattern, exp = clause
+      in
+      begin match pattern with
+        | Or (p1, p2) ->
+          eval_clauses constraints ((p1,exp)::(p2,exp)::tl) 
+        | Constructor (k, plist) ->
+          let pi = {idx=idx; op=piop_of_constructor k} in
+          let notpi = {pi with op=negate_piop pi.op} in 
+          let constraints' = (notpi :: constraints) in
+          begin match k with
+            | Int _ | String _ | Bool _ ->
+              assert (plist = []);
+              Leaf ([pi], exp)::(eval_clauses constraints' tl)
+            | Cons | Tuple -> assert (plist <> []);
+              let flat_children = plist |>
+                                  List.mapi (fun i pattern -> eval_clauses ~idx:i [] [(pattern, exp)]) |>
+                                  BatList.n_cartesian_product |>
+                                  List.map (fun n -> Conjunction n)
+              in
+              flat_children @ eval_clauses constraints tl
+            | Variant _ when plist <> [] ->
+              let children = clauses |>
+                          List.filter_map (shares_kst k) |>
+                          eval_clauses [] in
+              let brothers = clauses |>
+                             List.filter (fun c ->
+                                 shares_kst ~wildcard_to_none:true k c |> Option.is_none) |>
```

Comment:
Maybe `wildcard_to_none:false` could be rephrased as `include_wildcard: true`? We ask whether another pattern has "the same constructor", and the parameters say whether the notion of "same constructor" includes wildcards or not.

Reply:
That was the original name, but I named it `wildcard_to_none` because the result is piped to Option.is_none so that would make the meaning of the flag more explicit

Reply:
Now the parameter is a named one instead of optional